org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/CompletionEngine.java

```diff
@@ -11778,29 +11785,16 @@ private void findTypesAndPackages(char[] token, Scope scope, boolean proposeBase
 			}
 		}
 
+		checkCancel();
 		if (proposeConstructor && !isEmptyPrefix) {
-
-			checkCancel();
-
 			findTypesFromImports(token, scope, proposeType, typesFound);
 		} else if(proposeType) {
-
-			checkCancel();
-
 			findTypesFromStaticImports(token, scope, proposeAllMemberTypes, typesFound);
 		}
 
-		if (proposeConstructor) {
-
-			checkCancel();
-
-			findTypesFromExpectedTypes(token, scope, typesFound, proposeType, proposeConstructor);
-		}
-
+		checkCancel();
 		if (isEmptyPrefix && !this.assistNodeIsAnnotation) {
-			if (!proposeConstructor) {
-				findTypesFromExpectedTypes(token, scope, typesFound, proposeType, proposeConstructor);
```

Comment:
If I understand the changes up until this point, this line and the previous line 11797 were moved above, to line 11643?

Could you explain what that move achieves?

Reply:
A: The `findTypesFromExpectedTypes` finds current source type as a expected type. and since its a array type we create a completion even though is the same source type `(!inSameUnit || (inSameUnit && refBinding.isMemberType()) || (inSameUnit && isArrayCompletion)` at 12286

B: At the same time at 11666 where we go through the current types in the enclosing source type we find the same. 

Now in previous code, We find the current source type when running the code block at B:, And then we also found the same time at A:, but it was too late to change it since we have already added the type completion with dimension 0 insteand of correct array dimension.

By reordering this A: and B: to run A: first and B: after that, we could stop processing the current type if it was already handled in the `findTypesFromExpectedTypes` as part of array expected type. The exclusion happens at line 11715.